sdk/core/core-util/src/nonExhaustiveUnion.ts

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ * Represents a primitive JavaScript type string, number, bigint and symbol.
+ */
+export type Primitive<T> = T extends string
+  ? string
+  : T extends number
+  ? number
+  : T extends bigint
+  ? bigint
+  : T extends symbol
+  ? symbol
+  : unknown;
+
+/**
+ * This is a unique symbol that we declare. This symbol is used as a property key in our $NonExhaustive type.
+ * The use of a unique symbol guarantees that this property won't conflict with any other string-named properties that might exist.
+ */
+declare const __non_exhaustive: unique symbol;
+/**
+ * This is a mapped type that represents an object with a single optional property. The property key is the __non_exhaustive symbol we declared earlier, and the property value is of type never.
+ */
+export type $NonExhaustive = { [__non_exhaustive]?: never };
```

Comment:
Why we need to export `$NonExhaustive`?

Reply:
Because it is used by the NonExhaustive type, I tried to hide it but Api Extractor keeps yelling at me.

Reply:
Just curious why we add this property `[__non_exhaustive]`? Not empty {} like below:

```
export type NonExhaustive<T> = T | (Primitive<T> & {});
```